### PR TITLE
Feat/#102/users upload via csv

### PR DESCRIPTION
**Pull request by kadmit**
Task: https://github.com/keeleinstituut/tv-tolkevarav/issues/102

I implemented the functionality of uploading users via CSV. The solution includes 4 endpoints:

- Validation of the file. The endpoint returns a list of errors that were detected.
- Validation of the file row. The endpoint will be used to check changes after editing of incorrect row.
- Import the file. The endpoint will create institution users based on the data provided inside the file. In case the file has errors exception will be thrown.
- Endpoint for getting a list of institution users with filtering and sorting.

The task is not done fully because audit logs, tags, and vendor parts are not implemented yet.

### CODE REVIEW

File: application/app/Rules/UserFullNameRule.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+namespace App\Rules;
+
+use Closure;
+use Illuminate\Contracts\Validation\ValidationRule;
+
+class UserFullNameRule implements ValidationRule
+{
+    private const USER_NAME_REGEX = '/^[a-zõäöüšž\-]+(\s[a-zõäöüšž\-]+)$/iu';
```
Comment (DanBondarenko) [inline application/app/Rules/UserFullNameRule.php:10]: **[non-blocker]** Hmm, is it confirmed that we won’t to allow characters used in non-Estonian names? If this is confirmed with analysis, no problem, but I imagine there might probably be users with names such as "Radwan Dąbrowski-Żądło" or "Jānis Bērziņš".

File: application/database/factories/InstitutionUserFactory.php [hunk 2]
```diff
@@ -24,6 +24,7 @@ public function definition(): array
             'user_id' => User::factory(),
             'status' => InstitutionUserStatus::Created,
             'email' => $this->faker->email,
+            'phone' => $this->faker->numerify('+3725#######'),
```
Comment (DanBondarenko) [inline application/database/factories/InstitutionUserFactory.php:27]: **[compliment]** Nice!

File: application/database/migrations/2023_05_02_133231_add_phone_to_institution_users_table.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::table('institution_users', function (Blueprint $table) {
+            $table->string('phone')->nullable();
+        });
+    }
```
Comment (DanBondarenko) [inline application/database/migrations/2023_05_02_133231_add_phone_to_institution_users_table.php:17]: **[blocker]** While doing another task, I heard from Birgit that is phone numbers are required for users. With that in mind, shouldn’t the column be non-nullable?

File: application/app/Helpers/CsvReader.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    public function headers(): array
+    {
+        rewind($this->file);
+
+        return fgetcsv($this->file, 0, $this->separator);
+    }
+
+    public function rows(): Generator
+    {
+        $idx = 0;
```
Comment (DanBondarenko) [inline application/app/Helpers/CsvReader.php:25]: **[non-blocker]**
```suggestion
```

**Review by DanBondarenko — CHANGES_REQUESTED** [review]
Hi @kadmit! I found some minor thingies and a few issues. Sorry for not doing the review earlier – in retrospect I could have reused some code that you already wrote.

**Comment by kadmit** [comment]
@DanBondarenko thank you for the review! I fixed all of your suggestions, please take a look 🙂

**Comment by DanBondarenko** [comment]
> @DanBondarenko thank you for the review! I fixed all of your suggestions, please take a look 🙂

@kadmit Thank you for the changes and your patience! Looks great. Approving. :)

**Review by DanBondarenko — APPROVED** [review]

**Review by thenouan — APPROVED** [review]

**Pull request merged**